File: backend/app/schemas.py
"""Pydantic models for the API and internal data flow."""
from typing import Literal, Optional

from pydantic import BaseModel, Field


# --- API request ---
class ExecuteRequest(BaseModel):
    prompt: str = Field(..., min_length=1, max_length=8000)
    # Note: tenant_id / user_id are intentionally NOT request fields. They are derived
    # server-side from the verified access-token claims (see app/auth.py) so they cannot
    # be spoofed by the client.


# --- Internal: result of the single structured Claude call ---
class GenerationResult(BaseModel):
    should_execute: bool
    language: Optional[str] = None
    code: Optional[str] = None
    message: Optional[str] = None


# --- Internal: result of running code in a sandbox ---
class SandboxResult(BaseModel):
    stdout: str
    stderr: str
    exit_code: int
    duration_ms: int
    timed_out: bool


# --- API responses (discriminated by `type`) ---
class MessageResponse(BaseModel):
    """Returned when code generation does not make sense for the prompt."""

    type: Literal["message"] = "message"
    message: str


class ResultResponse(BaseModel):
    """Returned when code was generated and executed in the sandbox."""

    type: Literal["result"] = "result"
    language: str
    code: str
    stdout: str
    stderr: str
    exit_code: int
    duration_ms: int
    timed_out: bool

File: backend/src/schemas.ts
/** Zod request schema, internal result types, and response builders for the API. */
import { z } from "zod";

// --- API request ---
// tenant_id / user_id are intentionally NOT request fields; they are derived
// server-side from the verified token (see auth.ts) so they cannot be spoofed.
export const ExecuteRequest = z.object({
  prompt: z.string().min(1).max(8000),
  session_id: z.string().uuid().optional(), // continue an existing session; omit to start one
});
export type ExecuteRequest = z.infer<typeof ExecuteRequest>;

// --- Internal: result of the single structured Claude call ---
export interface GenerationResult {
  shouldExecute: boolean;
  language: string | null;
  code: string | null;
  message: string | null;
}

// --- Internal: result of running code in a sandbox ---
export interface SandboxResult {
  stdout: string;
  stderr: string;
  exitCode: number;
  durationMs: number;
  timedOut: boolean;
}

// --- API responses (discriminated by `type`, snake_case on the wire) ---
// Optional persistence trailer: `session_id`/`run_id` appear only when the run was persisted
// for an authenticated caller (both present, or neither). Anonymous / history-off omits them,
// keeping the response byte-identical to the pre-history contract.
type Persisted = { sessionId?: string; runId?: string };
const persistedWire = (p?: Persisted) =>
  p?.sessionId && p?.runId ? { session_id: p.sessionId, run_id: p.runId } : {};

export function messageResponse(message: string, p?: Persisted) {
  return { type: "message" as const, message, ...persistedWire(p) };
}

export function resultResponse(language: string, code: string, r: SandboxResult, p?: Persisted) {
  return {
    type: "result" as const,
    language,
    code,
    stdout: r.stdout,
    stderr: r.stderr,
    exit_code: r.exitCode,
    duration_ms: r.durationMs,
    timed_out: r.timedOut,
    ...persistedWire(p),
  };
}
